src/main/scala/cromwell/binding/WdlNamespace.scala

```diff
@@ -2,19 +2,109 @@ package cromwell.binding
 
 import java.io.File
 
-import cromwell.binding.WdlNamespace.ImportResolver
-import cromwell.binding.command._
 import cromwell.binding.types._
 import cromwell.binding.values.WdlValue
 import cromwell.parser.{AstTools, WdlParser}
 import cromwell.parser.WdlParser._
 import cromwell.util.FileUtil
-import AstTools.EnhancedAstNode
+import cromwell.parser.AstTools.{AstNodeName, EnhancedAstNode, EnhancedAstSeq}
 import scala.collection.JavaConverters._
 import scala.util.{Failure, Success, Try}
 import scala.language.postfixOps
 
 /**
+ * Define WdlNamespace as a sum type w/ two states - one containing a local workflow and one without.
+ * The latter is a valid state for a WDL file, however only the former can be requested to be run, so
+ * any constructs (e.g. WorkflowManagerActor) expecting to run a workflow should only take the `NamespaceWithWorkflow`
+ */
+sealed trait WdlNamespace extends WdlValue {
+  final val wdlType = WdlNamespaceType
+
+  def ast: Ast // FIXME: I think this is only used by the syntax highlighting, can it go away once we're built?
+  def importedAs: Option[String] // Used when imported with `as` 
+  def imports: Seq[Import] // FIXME: Change to Set?
+  def namespaces: Seq[WdlNamespace] // FIXME: Change to Set? FIXME: Rename to importedNamespaces?
+  def tasks: Seq[Task] // FIXME: Change to Set?
+  def importedWorkflows: Seq[Workflow] // FIXME: Change to Set?
+  def terminalMap: Map[Terminal, WdlSource]
+  def allWorkflows: Set[Workflow]
+
+  // Convenience method for findTask in the context of this namespace
+  def findTask(name: String): Option[Task] = WdlNamespace.findTask(name, namespaces, tasks)
+
+  override def toRawString = ???
+}
+
+/**
+ * A valid Namespace which doesn't have a locally defined Workflow. This should pass any validity checking but is not
+ * directly runnable by `WorkflowManagerActor`
+ */
```

Comment:
I just hope you can understand that your preference for lots and lots of ADTs is a subjective preference and not an objective measure of quality.  I get the feeling that when you say things like "ADTs are better because blah blah..." that you think that they're ALWAYS better under every circumstance.  Believe it or not, there are people out write very high quality code without going crazy with ADTs.  Hermes and the Python implementation of WDL are both very good and I am sparse with my use of ADTs.  I use them when its appropriate, but I also consider that things like code complexity and code size are also important and shouldn't be sacrificed unconditionally for more and more ADTs.  Sometimes a simple function or an if statement is better than a new type, in my opinion.